// lib/gcm.js
var user = require('config/models');
var room = require('lib/room');
var request = require('request');

exports.updateToken = function (token, gcmToken, callback) {
    // console.log("the id is " + id + " gcm_token is " + gcm_token);
    user.find({ token: token }, function (err, users) {
        if (users.length != 0) {
            var mUser = users[0];
            mUser.gcm_token = gcmToken;
            mUser.save();

            callback({ 'response': "Update gcm token success.", 'res': true });
        } else {
            callback({ 'response': "Error while updating gcm token.", 'res': false });
        }
    });
}

exports.callAllFriends = function (token, des, callback) {
    user.find({ token: token }, function (err, users) {
        if (users.length == 0) {
            callback({ 'response': "Error while calling all friends", 'res': false });
            return;
        }
        var mUser = users[0];
        if (mUser.friends.length == 0) {
            callback({ 'response': "Have no friend", 'res': false });
            return;
        }

        var roomCreated = false;
        var reporter = 0;

        for (var i = 0; i < mUser.friends.length; i++) {
            var friend = JSON.parse(mUser.friends[i]);
            user.find({ token: friend.token }, function (err, fusers) {
                if (fusers.length == 0) {
                    console.log("Cannot find " + friend.username + " and ignore.")
                } else {
                    var callee = fusers[0];
                    if (callee.gcm_token) {
                        console.log("Call " + callee.username);
                        callOneFriend(mUser.username, callee.gcm_token, token, des);
                        if (!roomCreated) {
                            roomCreated = true;
                            room.createRoom(token, des, "private", function(found) {
                                console.log(found);
                            });
                        }
                    } else {
                        console.log("Cannot call " + callee.username);
                    }
                }

                reporter += 1;
                if (reporter == mUser.friends.length) {
                    if (roomCreated) {
                        callback({ 'response': "Sent notifications to friends", 'res': true });
                    } else {
                        callback({ 'response': "Did not send notifications to any friend", 'res': false });
                    }
                }
            });
        }
    });
}

exports.callFriendsById = function (token, friends, des, callback) {
    user.find({token: token}, function(err, users) {
        if (users.length == 0) {
            callback({'response': "Error while calling friends.", 'res': false});
            return;
        }

        var mUser = users[0];
        if (mUser.friends.length == 0) {
            callback({ 'response': "Have no friend", 'res': false });
            return;
        }

        var roomCreated = false;
        var reporter = 0;

        for (var i = 0; i < friends.length; i++) {
            user.find({token: friends[i]}, function(err, fusers) {
                if (fusers.length == 0) {
                    console.log("Cannot find and ignore " + friends[i]);
                } else {
                    var callee = fusers[0];
                    var query = JSON.stringify({
                        'username': callee.username,
                        'token': callee.token
                    });
                    if (mUser.friends.indexOf(query) == -1) {
                        console.log(callee.username + " is not a friend!");
                    } else {
                        if (callee.gcm_token) {
                            console.log("Call " + callee.username);
                            callOneFriend(mUser.username, callee.gcm_token, token, des);
                            if (!roomCreated) {
                                roomCreated = true;
                                room.createRoom(token, des, "private", function(found) {
                                    console.log(found);
                                });
                            }
                        } else {
                             console.log("Cannot call " + callee.username);
                        }
                    }
                }

                reporter += 1;
                if (reporter == mUser.friends.length) {
                    if (roomCreated) {
                        callback({ 'response': "Sent notifications to friends", 'res': true });
                    } else {
                        callback({ 'response': "Did not send notifications to any friend", 'res': false });
                    }
                }
            });
        }
    });
}

var callOneFriend = function (username, friendGcmToken, roomId, des) {
    request({
        url: 'https://gcm-http.googleapis.com/gcm/send',
        headers: {
            'Content-Type': 'application/json',
            Authorization: 'key=AIzaSyBfR312em44wB7NPmVXwAJouVjj9Rz7jes'
        },
        method: 'POST',
        json: true,
        body: {
            "data": {
                sender: username,
                room_id: roomId,
                des: des,
                time: (new Date()).getTime
            },
            "to": friendGcmToken
        }
    }, function () { });
};
