// lib/gcm.js
var user = require('config/models');
var room = require('lib/room');
var request = require('request');

exports.update_token = function (id, gcm_token, callback) {
    // console.log("the id is " + id + " gcm_token is " + gcm_token);
    user.find({ token: id }, function (err, users) {
        if (users.length != 0) {
            var mUser = users[0];
            mUser.gcm_token = gcm_token;
            mUser.save();

            callback({ 'response': "Update Gcm Token Success", 'res': true });

        } else {
            callback({ 'response': "Error while updating gcm token", 'res': false });
        }
    });
}

// exports.call_friends = function (id, roomId, isNavigation, callback) {
//     user.find({ token: id }, function (err, users) {
//         if (users.length != 0) {
//             var mUser = users[0];
            
//             user.find({ username: { $in: mUser.friends } }, function(err, friends) {
//                 if (friends.length != 0) {
//                     callback({ 'response': "Sending notifications to friends", 'res': true });
//                     // console.log(friends);
//                     for (var i = 0; i < friends.length; i++) {
//                         if (friends[i].gcm_token) {
//                             call_friend(mUser.username, friends[i].gcm_token, roomId, isNavigation);
//                         } else {
//                             console.log("Cannot call " + friends[i].username);
//                         }
//                     }
//                 } else {
//                     callback({ 'response': "User has no friends", 'res': false });
//                 }
//             });
//         } else {
//             callback({ 'response': "Cannot find the user", 'res': false });
//         }
//     });
// }

exports.call_friends_by_id = function (id, friends, des, callback) {
    user.find({token: id}, function(err, users) {
        if (users.length != 0) {
            var mUser = users[0];
            var called = true;
            for (var i = 0; i < friends.length; i++) {
                user.find({token: friends[i]}, function(err, users2) {
                    if (users2.length != 0) {
                        var callee = users2[0];
                        var query = JSON.stringify({
                            'username': callee.username,
                            'id': callee.token
                        });
                        if (mUser.friends.indexOf(query) != -1) {
                            if (callee.gcm_token) {
                                console.log("Call " + callee.username);
                                var rateFloat = Math.round((mUser.rateTotal / mUser.rateNum) * 10) / 10;
                                call_friend(mUser.username, callee.gcm_token, id, des, rateFloat);
                                called = true;
                            } else {
                                console.log("Cannot call " + callee.username);
                            }
                        } else {
                            console.log(callee.username + " is not a friend!");
                        }
                    } else {
                        console.log("Cannot find and ignore " + friends[i]);
                    }
                });
            }
            if (called) {
                room.create_room(id, des, "private", function(found) {
                    console.log(found);
                });
                callback({ 'response': "Sending notifications to selected friends", 'res': true });
            } else {
                callback({ 'response': "Didnot send notifications to any friend", 'res': false });
            }
        } else {
            callback({'response': "Cannot find user while call friends", 'res': false});
        }
    });
}

var call_friend = function (username, friend_token, room_id, des, rate) {
    request({
        url: 'https://gcm-http.googleapis.com/gcm/send',
        headers: {
            'Content-Type': 'application/json',
            Authorization: 'key=AIzaSyBfR312em44wB7NPmVXwAJouVjj9Rz7jes'
        },
        method: 'POST',
        json: true,
        body: {
            "data": {
                sender: username,
                room_id: room_id,
                des: des,
                rate: rate,
                time: (new Date()).getTime
            },
            "to": friend_token
        }
    }, function () { });
};
