// lib/room.js
var User = require('config/models'); 
var Rate = require('lib/rate');

var apiKey = "45626032";
var apiSecret = "cefd5abf40619617780c2398fec72ace2ca41cb5";
var OpenTok = require('opentok');
var opentok  = new OpenTok(apiKey, apiSecret);

var room_list = {};

function keepAlive(resp) {
    if (resp) {
        resp.write(":\n");
        resp.keepAliveTimer = setTimeout(arguments.callee, 30000, resp);
    }
};
function kill(resp) {
    if (resp) {
        clearTimeout(resp.keepAliveTimer);
        resp.end();
    }
};

exports.createRoom = function(token, des, mode, callback) {
    // blind create room
    User.find({token: token}, function(err, users) {
        if (users.length != 0) {
            var room = room_list[token];
            if (room) {
                close(room);
                console.log("blind user " + token + " rejoin");
            } else {
                console.log("blind user " + token + " join");
            }
            var rateFloat = Math.round((users[0].rateTotal / users[0].rateNum) * 10) / 10;
            room_list[token] = {
                "private": mode === "private",
                "blind_id": token,
                "blind_name": users[0].username,
                "blind_res": null,
                "des": des,
                "rate": rateFloat,
                "helpers": {}
            };
            if (users[0].addresses && des.startsWith("navigation:")) {
                users[0].addresses.push(des);
                users[0].save();
            }
            callback({'response': "Create room success", 'res': true});
        } else {
            callback({'response': "Error while creating room", 'res': false});
        }
    });
};

exports.blind_keep_alive = function (id, res) {
    // blind offer res
    User.find({ token: id }, function (err, users) {
        if (users.length != 0) {
            var room = room_list[id];
            if (room) {
                var headers = {
                    "Content-Type": "text/event-stream",
                    "Cache-Control": "no-cache, no-store",
                    "Pragma": "no-cache",
                    "Expires": "0"
                };
                res.writeHead(200, headers);

                if (room.blind_res) {
                    kill(room.blind_res);
                    console.log("blind user " + id + " update res");
                } else {
                    console.log("blind user " + id + " offer res");
                }
                room.blind_res = res;
                keepAlive(res);

                refresh(room);
            } else {
                res.send("error:Cannot find room while blind offer res");
                console.log("Cannot find room while blind offer res");
            }
        } else {
            res.send("error:Cannot find user while blind offer res");
            console.log("Cannot find user while blind offer res");
        }
    });
};

exports.helperJoinRoom = function(token, roomId, res, callback) {
    // helper join room in list
    User.find({token: token}, function(err, users) {
        if (users.length == 0) {
            callback({'response': "Error while helper join", 'res': false});
        } else {
            var user = users[0];
            var room = room_list[roomId];
            if (!room) {
                callback({'response': "Cannot find room while helper join", 'res': false});
            } else {
                if (room.helpers[token]) {
                    kill(room.helpers[token].res);
                    delete room.helpers[token];
                    console.log("helper " + token + " rejoin room " + roomId);
                } else {
                    console.log("helper " + token + " join room " + roomId);
                }
                var rate = Math.round((user.rateTotal / user.rateNum) * 10) / 10;
                room.helpers[token] = {
                    "token": token,
                    "username": user.username,
                    "res": null,
                    "rate": rate
                };
                refresh(room);
                
                callback({'response': "Helper join success", 'res': true});
            }
        }
    });
};

exports.helper_keep_alive = function(id, room_id, res) {
    // helper offer res
    User.find({token: id}, function(err, users) {
        if (users.length != 0) {
            var user = users[0];
            var room = room_list[room_id];
            if (room) {
                var helper = room.helpers[id];
                if (helper) {
                    var headers = {
                        "Content-Type": "text/event-stream",
                        "Cache-Control": "no-cache, no-store",
                        "Pragma": "no-cache",
                        "Expires": "0"
                    };
                    res.writeHead(200, headers);
                    
                    if (helper.res) {
                        kill(helper.res);
                        console.log("helper " + id + " update res");
                    } else {
                        console.log("helper " + id + " offer res");
                    }
                    helper.res = res;
                    keepAlive(res);
                } else {
                    res.send("error:Not in the list while helper offer res");
                }
            } else {
                res.send("error:Room is closed while helper offer res");
            }
        } else {
            res.send("error:Cannot find user while helper offer res");
        }
    });
};

exports.leaveRoom = function(token, roomId, callback) {
    // helper leave room
    User.find({token: token}, function(err, users) {
        if (users.length == 0) {
            callback({'response': "Error while leaving room", 'res': false});
        } else {
            var user = users[0];
            var room = room_list[roomId];
            if (!room) {
                callback({'response': "Cannot find room while leaving", 'res': false});
            } else {
                if (room.helpers[token]) {
                    kill(room.helpers[token].res);
                    delete room.helpers[token];
                    refresh(room);
                    callback({'response': "Leave room success", 'res': true});
                } else {
                    callback({'response': "Not in this room", 'res': false});
                }
            }
        }
    });
};

exports.select = function(blind_id, helper_id, callback) {
    User.find({token: blind_id}, function(err, users) {
        if (users.length != 0) {
            var blind = users[0];
            var room = room_list[blind_id];
            if (room) {
                var helper = room.helpers[helper_id];
                if (helper && helper.res) {
                    /* Create session and token */
                    opentok.createSession({mediaMode:"routed"}, function(error, session) {
                        if (error) {
                            console.log("Error creating opentok session:", error);
                            if (room.blind_res) {
                                room.blind_res.write("error:Error creating opentok session");
                            }
                        } else {
                            var sessionId = session.sessionId;
                            console.log("Session ID: " + sessionId);
                            
                            var videoToken = opentok.generateToken(sessionId);
                            console.log("Token: " + videoToken);
                            
                            helper.res.write("select:" + JSON.stringify({
                                'select': true,
                                'session': sessionId,
                                'token': videoToken
                            }) + "\n");
                            kill(helper.res);
                            delete room.helpers[helper_id];
                            close(room);
                            callback({
                                'response': "Select success, start video", 
                                'res': true,
                                'session': sessionId,
                                'token': videoToken
                            });
                            
                        }
                    });
                    
                } else {
                    refresh(room);
                    callback({'response': "Helper not in list", 'res': false});
                }
            } else {
                callback({'response': "Cannot find room while select", 'res': false});
            }
        } else {
            callback({'response': "Cannot find user while select", 'res': false});
        }
    });
};

exports.deleteRoom = function(token, callback) {
    // blind close room
    User.find({token: token}, function(err, users) {
        if (users.length != 0) {
            var room = room_list[token];
            if (room) {
                close(room);
                callback({'response': "Close room success", 'res': true});
            } else {
                callback({'response': "Cannot find room", 'res': false});
            }
        } else {
            callback({'response': "Error while close room", 'res': false});
        }
    });
};

exports.getRooms = function(token, callback) {
    User.find({token: token}, function(err, users) {
        if (users.length != 0) {
            var rooms = [];
            var count = 0;
            for (var room_id in room_list) {
                if (count >= 20) break;
                var room = room_list[room_id];
                if (room.private) continue;
                
                count++;
                rooms.push({
                    "username": room.blind_name,
                    "room_id": room.blind_id,
                    "des": room.des,
                    "rate": room.rate
                });
            }
            callback({
                'response': "Get rooms success", 
                'res': true,
                'size': count,
                'list': JSON.stringify(rooms)
            });
        } else {
            callback({'response': "Error while close room", 'res': false});
        }
    });
};

var refresh = function(room) {
    if (!room.blind_res) return;
    console.log("Refresh room: " + room.blind_id);
    var namelist = [];
    for (var helper_id in room.helpers) {
        var helper = room.helpers[helper_id];
        namelist.push({
            "username": helper.username,
            "token": helper.token,
            "rate": helper.rate
        });
    }
    room.blind_res.write("refresh:" + JSON.stringify(namelist) + "\n");
};

var close = function(room) {
    kill(room.blind_res);
    for (var helper_id in room.helpers) {
        if (room.helpers[helper_id].res) {
            room.helpers[helper_id].res.write("select:" + JSON.stringify({
                'select': false
            }) + "\n");
            kill(room.helpers[helper_id].res);
        }
    }
    delete room_list[room.blind_id];
};
